plugins/UM3NetworkPrinting/src/Cloud/CloudOutputDevice.py

```diff
@@ -0,0 +1,424 @@
+class CloudOutputDevice(NetworkedPrinterOutputDevice):
+
+    # The interval with which the remote clusters are checked
+    CHECK_CLUSTER_INTERVAL = 20.0  # seconds
+
+    # Signal triggered when the print jobs in the queue were changed.
+    printJobsChanged = pyqtSignal()
+
+    # Signal triggered when the selected printer in the UI should be changed.
+    activePrinterChanged = pyqtSignal()
+
+    # Notify can only use signals that are defined by the class that they are in, not inherited ones.
+    # Therefore we create a private signal used to trigger the printersChanged signal.
+    _clusterPrintersChanged = pyqtSignal()
+
+    ## Creates a new cloud output device
+    #  \param api_client: The client that will run the API calls
+    #  \param cluster: The device response received from the cloud API.
+    #  \param parent: The optional parent of this output device.
+    def __init__(self, api_client: CloudApiClient, cluster: CloudClusterResponse, parent: QObject = None) -> None:
+        super().__init__(device_id = cluster.cluster_id, address = "",
+                         connection_type = ConnectionType.CloudConnection, properties = {}, parent = parent)
+        self._api = api_client
+        self._cluster = cluster
+
+        self._setInterfaceElements()
+
+        self._account = api_client.account
+
+        # We use the Cura Connect monitor tab to get most functionality right away.
+        self._monitor_view_qml_path = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                                                   "../../resources/qml/MonitorStage.qml")
+
+        # Trigger the printersChanged signal when the private signal is triggered.
+        self.printersChanged.connect(self._clusterPrintersChanged)
+
+        # We keep track of which printer is visible in the monitor page.
+        self._active_printer = None  # type: Optional[PrinterOutputModel]
+
+        # Properties to populate later on with received cloud data.
+        self._print_jobs = []  # type: List[UM3PrintJobOutputModel]
+        self._number_of_extruders = 2  # All networked printers are dual-extrusion Ultimaker machines.
+
+        # We only allow a single upload at a time.
+        self._progress = CloudProgressMessage()
+
+        # Keep server string of the last generated time to avoid updating models more than once for the same response
+        self._received_printers = None  # type: Optional[List[CloudClusterPrinterStatus]]
+        self._received_print_jobs = None  # type: Optional[List[CloudClusterPrintJobStatus]]
+
+        # A set of the user's job IDs that have finished
+        self._finished_jobs = set()  # type: Set[str]
+
+        # Reference to the uploaded print job / mesh
+        self._tool_path = None  # type: Optional[bytes]
+        self._uploaded_print_job = None  # type: Optional[CloudPrintJobResponse]
+
+    ## Connects this device.
+    def connect(self) -> None:
+        if self.isConnected():
```

Comment:
Isn't this something that should live in the base class? This isn't just specific to the CloudOutputDevice.

Reply:
Well, that would change behaviour for other output devices and I'm not sure of the impact of that.